open_lm/main.py

```diff
@@ -730,11 +714,17 @@ def main(args):
             break
 
         epoch = epoch + 1
-        evaluation_metrics = None
-        if "val" in data and (epoch % args.val_frequency == 0 or done_training):
+        evaluation_metrics = []
+        if "val_list" in data and (epoch % args.val_frequency == 0 or done_training):
             # validate based on frequency and always validate the last checkpoint
             try:
-                evaluation_metrics = evaluate(model, data, epoch, args, writer)
+                evaluation_metrics = evaluate_loop(model, data["val_list"], epoch, args, writer)
+
+                if is_master(args):
+                    with fsspec.open(os.path.join(args.checkpoint_path, "results.jsonl"), "a") as f:
+                        f.write(json.dumps(evaluation_metrics))
```

Comment:
it might be good to do this as 1 write, since we are doing this over fsspec, there could be a network error and we are left with a corrupt jsonl or something (without the newline)

Reply:
sg